File: java/dfp-math/src/main/java/com/epam/deltix/dfpmath/NativeMathImplLoader.java
package com.epam.deltix.dfpmath;

import com.epam.deltix.utilities.*;

class NativeMathImplLoader {
    public static void load() {
        try {
            String version = NativeMathImpl.version;
            final boolean isSnapshot = version.endsWith("-SNAPSHOT");
            if (isSnapshot)
                version = version.substring(0, version.length() - "-SNAPSHOT".length());

            VariablesMapper varMapper = new VariablesMapper(NativeMathImplLoader.class, version);

            final String unpackEnvVarName = varMapper.getPackageLast().toUpperCase() + "_UNPACK_ROOT";
            String unpackPath = System.getenv(unpackEnvVarName);
            if (unpackPath == null)
                unpackPath = "$(TEMP)/$(PACKAGE)/$(VERSION)/$(ARCH)";

            ResourceLoader
                .from(NativeMathImplLoader.class, "resources_" + varMapper.getPackage().replace('.', '_') + '/' + varMapper.getOs() + '/' + varMapper.getArch() + "/*")
                .to(varMapper.substitute(unpackPath))
                .alwaysOverwrite(isSnapshot)
                .tryRandomFallbackSubDirectory(true)
                .load();
        } catch (final Throwable exception) {
            //exception.printStackTrace();
            throw new RuntimeException(exception);
        }
    }
}

File: csharp/EPAM.Deltix.DFPMath/NativeMathImplLoader.cs
using EPAM.Deltix.Utilities;
using System;
using System.IO;
using System.Runtime.InteropServices;
#if NET40
using System.Diagnostics;
using System.Linq;
using System.Text.RegularExpressions;
#endif

namespace EPAM.Deltix.DFPMath
{
	internal static class NativeMathImplLoader
	{
		internal static bool isLoaded = false;
		internal static object isLoadedLock = new object();
		internal static void Load()
		{
			if (isLoaded)
				return;
			lock (isLoadedLock)
			{
				if (isLoaded)
					return;

				var varMapper = new VariablesMapper(typeof(NativeMathImplLoader));

				var unpackEnvVarName = varMapper.PackageLast.ToUpperInvariant() + "_UNPACK_ROOT";
				var unpackPath = Environment.GetEnvironmentVariable(unpackEnvVarName);
				if (unpackPath == null)
					unpackPath = "$(TEMP)/$(PACKAGE)/$(VERSION)/$(ARCH)";

				var loader = ResourceLoader
					.From($"EPAM.Deltix.DFPMath.{varMapper.Os}.{varMapper.Arch}.*")
					.To(varMapper.Substitute(unpackPath))
					// .LowercasePathOnLinux(false)
					.TryRandomFallbackSubDirectory(true)
					.Load();

				isLoaded = true;
			}
		}
	}
}
